### PR TITLE
Enable retry on rate limit error

### PR DESCRIPTION
**Pull request by davidkyle**
The bulk of the change is building an informative error message from the OpenAI headers in the response. 

Non-issue as this is part of the general OpenAI text embedding model work

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java [hunk 1]
```diff
@@ -61,4 +79,56 @@ void checkForFailureStatusCode(HttpRequestBase request, HttpResult result) throw
             throw new RetryException(false, buildError(UNSUCCESSFUL, request, result));
         }
     }
+
+    static ElasticsearchStatusException buildRateLimitError(String baseMessage, HttpRequestBase request, HttpResult result) {
+        var response = result.response();
+        int statusCode = result.response().getStatusLine().getStatusCode();
+        var tokenLimit = getFirstHeader(response.getHeaders(TOKENS_LIMIT));
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java:86]: Would it make this code simpler to use `getFirstHeader()` defined in `HttpMessage` from https://hc.apache.org/httpcomponents-core-4.4.x/current/httpcore/apidocs/

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java [hunk 2]
```diff
@@ -61,4 +79,56 @@ void checkForFailureStatusCode(HttpRequestBase request, HttpResult result) throw
...
+        var remainingTokens = getFirstHeader(response.getHeaders(REMAINING_TOKENS));
+        var requestLimit = getFirstHeader(response.getHeaders(REQUESTS_LIMIT));
+        var remainingRequests = getFirstHeader(response.getHeaders(REMAINING_REQUESTS));
+
+        StringBuilder sb = new StringBuilder();
+        if (tokenLimit != null) {
```
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java:92]: Both seem fine to me but your suggestion here to use the `unknown` sounds good as it does reduce the number of if-blocks.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java [hunk 3]
```diff
@@ -19,9 +21,25 @@
...
+    // The maximum number of requests that are permitted before exhausting the rate limit.
+    static final String REQUESTS_LIMIT = "x-ratelimit-limit-requests";
+    // The maximum number of tokens that are permitted before exhausting the rate limit.
+    static final String TOKENS_LIMIT = "x-ratelimit-limit-tokens";
+    // The remaining number of requests that are permitted before exhausting the rate limit.
+    static final String REMAINING_REQUESTS = "x-ratelimit-remaining-requests";
+    // The remaining number of tokens that are permitted before exhausting the rate limit.
+    static final String REMAINING_TOKENS = "x-ratelimit-remaining-tokens";
+    // The time until the rate limit (based on requests) resets to its initial state.
+    static final String REQUESTS_LIMIT_RESET_TIME = "x-ratelimit-reset-requests";
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java:40]: Just to double check, we're intentionally not logging these right?

**Review by jonathan-buttner — APPROVED** [review]

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java (hunk 3 above)
Comment (davidkyle) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java]: No not used, I removed the constants

**Pull request merged**